Sources/Rendering/Core/Texture/index.js

```diff
@@ -100,20 +100,155 @@ function vtkTexture(publicAPI, model) {
     const dimensionality = (width > 1) + (height > 1) + (depth > 1);
     return dimensionality;
   };
+
+  publicAPI.getImageData = () => {
+    if (!model.imageLoaded || publicAPI.getInputData()) return null;
+
+    if (model.jsImageData) {
+      return model.jsImageData();
+    }
+    if (model.canvas) {
+      const context = model.canvas.getContext('2d');
+      const imageData = context.getImageData(
+        0,
+        0,
+        model.canvas.width,
+        model.canvas.height
+      );
+      return imageData;
+    }
+    if (model.image) {
+      const canvas = document.createElement('canvas');
+      canvas.width = model.image.width;
+      canvas.height = model.image.height;
+      const context = canvas.getContext('2d');
+      context.translate(0, canvas.height);
+      context.scale(1, -1);
+      context.drawImage(
+        model.image,
+        0,
+        0,
+        model.image.width,
+        model.image.height
+      );
+      const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+      return imageData;
+    }
+
+    return null;
+  };
 }
 
+// Use nativeArray instead of self
+const generateMipmaps = (nativeArray, width, height, level) => {
+  // TODO: FIX UNEVEN TEXTURE MIP GENERATION:
+  // When textures don't have standard ratios, higher mip levels
```

Comment:
Not an expert. I believe the standard approach is to filter first, then sample. (averaging the four pixels is a center point linear sampling) As written the level 0 texture is not being filtered prior to sampling and the other resolutions are being stored in their post filtered form. Might be worth allocating memory and filtering first.

Reply:
The way I have it set up right now, the level zero texture is the initial unchanged input data which should be unfiltered. Any new maps that are generated go through the linear down sample and then blur, and I haven't seen much difference by switching it the other way around

Reply:
you make a copy of the level 0, filter that and then sample it to get level 1. But this isn't my area of expertise. Do you have a reference that says to sample first then smooth? From wikipedia "A lowpass pyramid is made by smoothing the image with an appropriate smoothing filter and then subsampling the smoothed image, usually by a factor of 2 along each coordinate direction." But if you have a reference to sampling first then smoothing (or if someone who's background is image processing knows better speak up).

Reply:
And honestly I might even just skip the filtering if you don;t see much of a difference, especially if it takes a while

Reply:
I couldn't find anything really strong supporting one or the other after looking through a host of different resources. The method of mipmapping the environment map is an approximation to begin with, but a good approximation that saves time over actually sampling the map with progressively more scattered samples. I did some tests on sampling then blurring vs blurring then sampling, and the main difference is that the texture gets blurred faster when it's sampled first vs blurred first. The other difference is that it over doubles the time required to compute the mipmaps, since the base texture alone has more data than all the other mipmaps combined. Additionally, less blur may result in needing to use a kernel that's larger than 3x3 which would slow it down further. I'm fine with either, but from a speed and ease perspective sampling-then-blurring makes more sense.